integration/instance.go

```diff
@@ -0,0 +1,392 @@
+	longTimeout = 30 // seconds
+
+	// convergedTimeout is the number of seconds we wait for our instance to
+	// remain unchanged to be considered as converged.
+	convergedTimeout = 5 // seconds
+
+	// dirMode is the the mode used when making directories.
+	dirMode = 0755
+
+	// fileMode is the the mode used when making files.
+	fileMode = 0644
+)
+
+// Instance represents a single running mgmt instance. It is a building block
+// that can be used to run standalone tests, or combined to run clustered tests.
+type Instance struct {
+	// Hostname is a unique identifier for this instance.
+	Hostname string
+
+	// Preserve prevents the runtime output from being explicitly deleted.
+	// This is helpful for running analysis or tests on the output.
+	Preserve bool
+
+	// Debug enables more verbosity.
+	Debug bool
+
+	// dir is the directory where all files will be written under.
+	dir string
+
+	tmpPrefixDirectory   string
+	testRootDirectory    string
+	convergedStatusFile  string
+	convergedStatusIndex int
+
+	cmd *exec.Cmd
+
+	clientURL string // set when launched with run
+	serverURL string
+}
+
+// Init runs some initialization for this instance. It errors if the struct was
+// populated in an invalid way, or if it can't initialize correctly.
+func (obj *Instance) Init() error {
+	if obj.Hostname == "" {
+		return fmt.Errorf("must specify a hostname")
+	}
+
+	// create temporary directory to use during testing
+	if obj.dir == "" {
+		var err error
+		obj.dir, err = ioutil.TempDir("", fmt.Sprintf("mgmt-integration-%s-", obj.Hostname))
+		if err != nil {
+			return errwrap.Wrapf(err, "can't create temporary directory")
+		}
+	}
+
+	tmpPrefix := path.Join(obj.dir, PrefixDirectory)
+	if err := os.MkdirAll(tmpPrefix, dirMode); err != nil {
+		return errwrap.Wrapf(err, "can't create prefix directory")
+	}
+	obj.tmpPrefixDirectory = tmpPrefix
+
+	testRootDirectory := path.Join(obj.dir, RootDirectory)
+	if err := os.MkdirAll(testRootDirectory, dirMode); err != nil {
+		return errwrap.Wrapf(err, "can't create instance root directory")
+	}
+	obj.testRootDirectory = testRootDirectory
+
+	obj.convergedStatusFile = path.Join(obj.dir, ConvergedStatusFile)
+
+	return nil
+}
+
+// Close cleans up after we're done with this instance.
+func (obj *Instance) Close() error {
+	if !obj.Preserve {
+		if obj.dir == "" || obj.dir == "/" {
+			panic("obj.dir is set to a dangerous path")
+		}
+		if err := os.RemoveAll(obj.dir); err != nil { // dangerous ;)
+			return errwrap.Wrapf(err, "can't remove instance dir")
+		}
+	}
+	obj.Kill() // safety
+	return nil
+}
+
+// Run launches the instance. It returns an error if it was unable to launch.
+func (obj *Instance) Run(seeds []*Instance) error {
+	if obj.cmd != nil {
+		return fmt.Errorf("an instance is already running")
+	}
+
+	if len(seeds) == 0 {
+		// set so that Deploy can know where to connect
+		// also set so that we can allow others to find us and connect
+		obj.clientURL = "http://127.0.0.1:2379"
```

Comment:
I've had many times where race conditions causes tests to fail as old mgmt instances where occupying ports that where assumed to be free by the testsuite. A solution I was thinking about is using tcp dial on a 0 port to get a available port from the operating system, like: https://github.com/phayes/freeport/blob/master/freeport.go#L9.

Reply:
Ideally instead of tcp, unix domain sockets would be used during test to completely isolate tests and allow for parallel runs. This of course has to wait for the implementation.

Reply:
Agreed! I just did it this way since it's the only way I knew. For now it seems to not conflict, but if we experience issue in the future, we can change it.